### PR TITLE
Implement tuple return type

### PR DESCRIPTION
**Pull request by Smit-create**
#906

### CODE REVIEW

**Review by czgdp1807 — CHANGES_REQUESTED** [review]
The implementation seems okay to me. Can we add some good tests to verify that this feature works?

**Comment by Smit-create** [comment]
This change actually fails due to some error:
```py
def add1(a: i32, b: i32) -> tuple[i32, i32]:
    t: tuple[i32, i32]
    t = (a+1, b+1)
    return t

def f():
    i: i32 = 1
    j: i32 = 1
    t: tuple[i32, i32]
    t = add1(i, j)

f()
```

Error:
```
 tuple_api->tuple_deepcopy(value_tuple, target_tuple, type_code);
  File "/Users/thebigbool/repos/lpython/src/libasr/codegen/llvm_utils.cpp", line 339
    LFORTRAN_ASSERT(src->getType() == dest->getType());
AssertFailed: src->getType() == dest->getType()
```

Commenting out that line fails one more time here:
```
File "/Users/thebigbool/repos/lpython/src/libasr/codegen/llvm_utils.cpp", line 89
    return LLVM::CreateGEP(*builder, ds, idx_vec);
  File "/Users/thebigbool/repos/lpython/src/libasr/codegen/llvm_utils.cpp", line 23
    LFORTRAN_ASSERT(t->isPointerTy());
AssertFailed: t->isPointerTy()
```

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 1]
```diff
@@ -2472,6 +2472,25 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
...
+                        bool is_local_list = false;
+                        ASR::dimension_t* local_m_dims = nullptr;
+                        int local_n_dims = 0;
+                        int local_a_kind = -1;
+                        ASR::storage_typeType local_m_storage = ASR::storage_typeType::Default;
+                        llvm_el_types.push_back(get_type_from_ttype_t(asr_tuple->m_type[i], local_m_storage,
+                                                is_local_array_type, is_local_malloc_array_type,
+                                                is_local_list, local_m_dims, local_n_dims, local_a_kind));
+                    }
+                    return_type = tuple_api->get_tuple_type(type_code, llvm_el_types);
```
Comment (czgdp1807) [inline src/libasr/codegen/asr_to_llvm.cpp:2491]: I see. You need to return `return_type->getPointerTo()`. In simple words, the return type should be a pointer to struct and not the struct itself.

Comment (czgdp1807) [reply src/libasr/codegen/asr_to_llvm.cpp:2491]: Well as soon as the function call goes out of scope, `t`  inside `add1` in your case, pointer to `t` will be freed. So you may receive segmentation fault. But still try returning a pointer to struct and let’s see what happens.

Comment (czgdp1807) [reply src/libasr/codegen/asr_to_llvm.cpp:2491]: Or well you are anyways returning a value so just doing `builder->CreateStore(return_value, target)`.

Comment (Smit-create) [reply src/libasr/codegen/asr_to_llvm.cpp:2491]: <details>
<summary> LLVM code </summary>
<p>

```
; ModuleID = 'LFortran'
source_filename = "LFortran"

%tuple = type { i32, i32 }

define void @_lpython_main_program() {
.entry:
  call void @f()
  br label %return

return:                                           ; preds = %.entry
  ret void
}

define %tuple* @add1(i32* %a, i32* %b) {
.entry:
  %_lpython_return_variable = alloca %tuple, align 8
  %t = alloca %tuple, align 8
  %const_tuple = alloca %tuple, align 8
  %0 = add i32* %a, i32 1
  %1 = add i32* %b, i32 1
  %2 = getelementptr %tuple, %tuple* %const_tuple, i32 0, i32 0
  store i32* %0, i32* %2, align 8
  %3 = getelementptr %tuple, %tuple* %const_tuple, i32 0, i32 1
  store i32* %1, i32* %3, align 8
  %4 = getelementptr %tuple, %tuple* %const_tuple, i32 0, i32 0
  %5 = load i32, i32* %4, align 4
  %6 = getelementptr %tuple, %tuple* %t, i32 0, i32 0
  store i32 %5, i32* %6, align 4
  %7 = getelementptr %tuple, %tuple* %const_tuple, i32 0, i32 1
  %8 = load i32, i32* %7, align 4
  %9 = getelementptr %tuple, %tuple* %t, i32 0, i32 1
  store i32 %8, i32* %9, align 4
  %10 = getelementptr %tuple, %tuple* %t, i32 0, i32 0
  %11 = load i32, i32* %10, align 4
  %12 = getelementptr %tuple, %tuple* %_lpython_return_variable, i32 0, i32 0
  store i32 %11, i32* %12, align 4
  %13 = getelementptr %tuple, %tuple* %t, i32 0, i32 1
  %14 = load i32, i32* %13, align 4
  %15 = getelementptr %tuple, %tuple* %_lpython_return_variable, i32 0, i32 1
  store i32 %14, i32* %15, align 4
  br label %return

unreachable_after_return:                         ; No predecessors!
  br label %return

return:                                           ; preds = %unreachable_after_return, %.entry
  %16 = load %tuple, %tuple* %_lpython_return_variable, align 4
  ret %tuple %16
}

define void @f() {
.entry:
  %i = alloca i32, align 4
  %j = alloca i32, align 4
  %t = alloca %tuple, align 8
  store i32 1, i32* %i, align 4
  store i32 1, i32* %j, align 4
  %0 = call %tuple* @add1(i32* %i, i32* %j)
  %1 = getelementptr %tuple, %tuple* %0, i32 0, i32 0
  %2 = load i32, i32* %1, align 4
  %3 = getelementptr %tuple, %tuple* %t, i32 0, i32 0
  store i32 %2, i32* %3, align 4
  %4 = getelementptr %tuple, %tuple* %0, i32 0, i32 1
  %5 = load i32, i32* %4, align 4
  %6 = getelementptr %tuple, %tuple* %t, i32 0, i32 1
  store i32 %5, i32* %6, align 4
  br label %return

return:                                           ; preds = %.entry
  ret void
}

define i32 @main() {
.entry:
  call void @_lpython_main_program()
  ret i32 0
}
```
</p>
</details>

Error:
```
code generation error: asr_to_llvm: module failed verification. Error:
Both operands to a binary operator are not of the same type!
  %0 = add i32* %a, i32 1
Both operands to a binary operator are not of the same type!
  %1 = add i32* %b, i32 1
Stored value type does not match pointer operand type!
  store i32* %0, i32* %2, align 8
 i32Stored value type does not match pointer operand type!
  store i32* %1, i32* %3, align 8
 i32Function return type does not match operand type of return inst!
  ret %tuple %16
 %tuple*
```

Comment (czgdp1807) [reply src/libasr/codegen/asr_to_llvm.cpp:2491]: Well, I will push my changes tomorrow to your PR once I complete some work in LFortran. You can move forward with next item in your priority list.

**Comment by czgdp1807** [comment]
Closing in favour of https://github.com/lcompilers/lpython/pull/939

**Pull request closed without merging**